doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -395,35 +395,48 @@ conclusion.
 For backward compatibility, the notation :n:`in {+ @ident}` performs
 the conversion in hypotheses :n:`{+ @ident}`.
 
-.. tacn:: cbv {? @strategy_flag }
-          lazy {? @strategy_flag }
-   :name: cbv; lazy
-
-   These parameterized reduction tactics apply to any goal and perform
-   the normalization of the goal according to the specified flags. In
-   correspondence with the kinds of reduction considered in Coq namely
-   :math:`\beta` (reduction of functional application), :math:`\delta`
-   (unfolding of transparent constants, see :ref:`vernac-controlling-the-reduction-strategies`),
-   :math:`\iota` (reduction of
-   pattern matching over a constructed term, and unfolding of :g:`fix` and
-   :g:`cofix` expressions) and :math:`\zeta` (contraction of local definitions), the
-   flags are either ``beta``, ``delta``, ``match``, ``fix``, ``cofix``,
-   ``iota`` or ``zeta``. The ``iota`` flag is a shorthand for ``match``, ``fix``
-   and ``cofix``. The ``delta`` flag itself can be refined into
-   :n:`delta [ {+ @qualid} ]` or :n:`delta - [ {+ @qualid} ]`, restricting in the first
-   case the constants to unfold to the constants listed, and restricting in the
-   second case the constant to unfold to all but the ones explicitly mentioned.
+.. tacn:: lazy {? @strategy_flag } {? @occurrences }
+          cbv {? @strategy_flag } {? @occurrences }
+
+   .. todo PR: should this explanation move to where strategy_flag is defined?
+      or move that syntax definition here?
+
+   Normalize the goal as specified by the :n:`@strategy_flag`\s. The flags
+   are:
+
+   `beta` (:math:`\beta`)
+     reduction of functional application
+
+   `delta` (:math:`\delta`)
+     unfolding of transparent constants, see :ref:`vernac-controlling-the-reduction-strategies`
+
+   `iota` (:math:`\iota`)
+     reduction of pattern matching (`match`) over a constructed term and unfolding of :g:`fix` and
+     :g:`cofix` expressions.  Shorthand for `match fix cofix`
+
+   `zeta` (:math:`\zeta`)
+      contraction of local definitions
+
+   If no :n:`@strategy_flag` is given, the flags `beta delta iota zeta` are used.
```

Comment:
I used more words than you suggested.

- `*_flag` seems like a weak choice for the nonterminal names here.  "Flag" relates more to syntax; we should go for the meaning.  How about `*_conv`?
- `strategy` is one of those words you could use almost anywhere in the grammar.  What's a better name? What are the essential differences between `strategy_flag`, `red_flag` and `delta_flag`?
- It's odd for `strategy_flag` to support `delta_flag` with 2 different syntaxes: `delta_flag` AND `"delta" OPT delta_flag`
- Is there a good reason for some commands to only take `delta_flag`?  Consistency would be nice if it makes sense.

Reply:
Also, what does a `@red_flag` of `"delta"` mean without the the `@delta_flag`?  (Or with it, for that matter.)

Reply:
> * `strategy` is one of those words you could use almost anywhere in the grammar.  What's a better name? What are the essential differences between `strategy_flag`, `red_flag` and `delta_flag`?

`delta_flag` controls delta-reduction. The difference between `red_flag` and `delta_flag` is that `red_flag` supports more than just controlling delta-reduction, and furthermore requires an explicit `delta` before the whitelist/blacklist for delta-reduction, so that it's clear that the list of constants is attached to a delta reduction.  `strategy_flag` allows the user to drop the explicit `delta` modifier and give only a whitelist/blacklist to mean "beta iota zeta" + delta with the whitelist/blacklist.

> * It's odd for `strategy_flag` to support `delta_flag` with 2 different syntaxes: `delta_flag` AND `"delta" OPT delta_flag`

`delta_flag` is a shorter way to write `beta iota zeta delta delta_flag`.  Note that `delta_flag` is not the same as `delta delta_flag`; the later performs *only* delta reduction.

> * Is there a good reason for some commands to only take `delta_flag`?  Consistency would be nice if it makes sense.

For some strategies, it does not make sense to perform them without beta-iota-zeta reduction.  I think `vm_compute` and `native_compute` could not be modified to support `strategy_flag`, while it would be quite nice for them to support `delta_flag`.  I believe all the others could be modified to support `strategy_flag`.

> Also, what does a `@red_flag` of `"delta"` mean without the the `@delta_flag`? (Or with it, for that matter.)

`@red_flag` of `delta` means "unfold all constants" (except for `cbn` where it means "don't forbid unfolding of any constants).  With `@delta_flag`, only the constants in the whitelist (or not in the blacklist, if the `-[references]` form is used) are unfolded.

Reply:
>`*_flag` seems like a weak choice for the nonterminal names here.  "Flag" relates more to syntax; we should go for the meaning.  How about `*_conv`?

I like `flag` because it clearly shows that this is a parameter that you can tweak for this tactic.

> `strategy` is one of those words you could use almost anywhere in the grammar.  What's a better name? What are the essential differences between `strategy_flag`, `red_flag` and `delta_flag`?

`strategy_flag` just gives you the choice between specifying all the possible `red_flags` (reduction flags) or to only specify how to tweak the `delta` reduction (if you decide to use the default reduction flags).

> It's odd for `strategy_flag` to support `delta_flag` with 2 different syntaxes: `delta_flag` AND `"delta" OPT delta_flag`

As suggested by my suggestion above, this is not the same thing: when you specify `delta` explicitly (and only `delta`), you are turning off the other reductions (beta, iota, zeta) whereas when you specify the `delta_flag` only then you use the default reductions (the four of them). It makes sense to have a way to specify this without having to type the mouthful `beta delta @delta_flag iota zeta`.

I think the rest of your questions should be obvious given my explanation above.

Reply:
I must disagree.  `_flag` and `strategy_flag` are not very descriptive.  `strategy_flag` and `delta_flag` both take lists of items, so they should be plural.  These would be much more descriptive:
- red_flag -> reduction
- strategy_flag -> reductions
- delta_flag -> delta_reductions
WDYT?

Reply:
> @JasonGross whitelist/blacklist for delta-reduction

whitelist/blacklist meaning what is declared as Opaque or Transparent?  I'd like to describe which setting (Opaque/Transparent) is the default for various types of definitions.

Reply:
> whitelist/blacklist meaning what is declared as Opaque or Transparent?

Roughly, yes.  But the list will not allow the tactic to bypass anything that is Opaque, it will only forbid unfolding of some constants.

> I'd like to describe which setting (Opaque/Transparent) is the default for various types of definitions

Everything ending in Defined or defined with := is Transparent.  Things defined with Qed are even more opaque than Opaque, in that even conversion and vm_compute/native_compute treat them like axioms (which have no body and thus cannot be unfolded).  Note that when an empty  `delta_flag` is passed (e.g., `cbv` or `cbv delta`), this is the same as an empty blacklist, i.e., everything transparent is unfolded.

Reply:
> I must disagree. `_flag` and `strategy_flag` are not very descriptive. `strategy_flag` and `delta_flag` both take lists of items, so they should be plural. These would be much more descriptive:
> 
> * red_flag -> reduction
> * strategy_flag -> reductions 
> * delta_flag -> delta_reductions
> WDYT?

Fine with me